AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/NpmPackageDownloader.kt

```
@@ -0,0 +1,149 @@
+
+import android.content.Context
+import com.fasterxml.jackson.core.JacksonException
+import com.fasterxml.jackson.databind.DeserializationFeature
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.ichi2.anki.CollectionHelper
+import com.ichi2.anki.R
+import com.ichi2.anki.UIUtils
+import com.ichi2.anki.jsaddons.NpmUtils.isvalidAnkiDroidAddon
+import com.ichi2.anki.web.HttpFetcher
+import com.ichi2.async.ProgressSenderAndCancelListener
+import com.ichi2.async.TaskDelegate
+import com.ichi2.async.TaskListener
+import com.ichi2.libanki.Collection
+import org.apache.commons.compress.archivers.ArchiveException
+import timber.log.Timber
+import java.io.File
+import java.io.IOException
+import java.net.URL
+import java.net.UnknownHostException
+
+class NpmPackageDownloader {
+
+    class DownloadAddon(private val mContext: Context, private val mNpmPackageName: String) : TaskDelegate<Void?, String?>() {
+
+        override fun task(col: Collection, collectionTask: ProgressSenderAndCancelListener<Void?>): String? {
+            return download()
+        }
+
+        /**
+         * Using Jackson the latest package.json for the addon fetched, then mapped to AddonModel
+         * For valid package it gets tarball url from mapped AddonModel,
+         * then downloads and extract to AnkiDroid/addons folder using {@code extractAndCopyAddonTgz} and toast with success message returned
+         *
+         * For invalid addon or for exception occurred, it returns message to respective to the errors from catch block
+         */
+        fun download(): String? {
+            try {
+                // mapping for json fetched from http://registry.npmjs.org/ankidroid-js-addon-.../latest
+                val mapper = ObjectMapper()
+                    .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+                val addonModel = mapper.readValue(URL(mContext.getString(R.string.npmjs_registry, mNpmPackageName)), AddonModel::class.java)
```

Do we want to fail on unknown values?